### PR TITLE
Rate limiter

### PR DESCRIPTION
**Pull request by Samuelhertzberg**
Added dynamic rate limiting. 
Now the clients no longer share the pool of requests per second. Instead each client has their own pool.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -76,7 +92,22 @@ const dostroy = (config) => async (req, res, next) => {
   const all = !config || Object.keys(config).length === 0
   const r = config && config.rudy ? config.rudy : RUDY_DEFAULT
   const rl = config && config.rateLimiting ? config.rateLimiting : RATELIMITING_DEFAULT
+  const dynamic = config && config.rateLimiting ? config.rateLimiting.useDynamicRateLimiting : USE_DYNAMIC_RATE_LIMITING_DEFAULT
+  const userActiveTimeout = dynamic && config.rateLimiting.userActiveTimeout ? config.rateLimiting.userActiveTimeout : USER_ACTIVE_TIMEOUT_DEFAULT
```
Comment (helenaalinder) [inline index.js]: We should somehow check that the user input is valid, so that we don't get unexpected errors or comparisons

File: index.js [hunk 2]
```diff
@@ -5,20 +5,25 @@ const RUDY_DEFAULT = false
...
+const USE_DYNAMIC_RATE_LIMITING_DEFAULT = false
+const USER_ACTIVE_TIMEOUT_DEFAULT = 30000
 
 const logSession = new Date().toISOString()
 const logStream = fs.createWriteStream('/tmp/express-requests-' + logSession + '.log', { flags: 'a' })
 
-const _interval = 3000 // milliseconds to check number of requests
-const _limit = 3 // limit for requests within interval
+const _interval = 10000 // milliseconds to check number of requests
+const _limit = 10 // limit for requests within interval
```
Comment (helenaalinder) [inline index.js]: 10 requests/10 seconds is really slow and difficult to match when testing, should maybe changes to 3 req/seconds or something.

File: index.js [hunk 3]
```diff
@@ -30,26 +35,37 @@ const rateLimiting = (req, res, next, logging) => {
...
 
   let requests = addressObject.requests
   const startRequestAt = addressObject.startRequestAt
   const diffSeconds = now.diff(startRequestAt)
-  if (diffSeconds < _interval && requests > _limit) {
-    addAddressToRequests(address, requests + 1, now)
+  const limitDenominator = (totalActiveUsers > 0 ? totalActiveUsers : 1) //1 if there is no dynamic rate limiting
+  const requestLimit = _limit/limitDenominator
+  if (diffSeconds < _interval && requests > requestLimit) {
+    console.log(address + ' broke the limit of ' + requestLimit + ' connections')
```
Comment (helenaalinder) [inline index.js]: Remove this printout maybe

File: index.js [hunk 4]
```diff
@@ -5,20 +5,25 @@ const RUDY_DEFAULT = false
...
+const USER_ACTIVE_TIMEOUT_DEFAULT = 30000
 
 const logSession = new Date().toISOString()
 const logStream = fs.createWriteStream('/tmp/express-requests-' + logSession + '.log', { flags: 'a' })
 
-const _interval = 3000 // milliseconds to check number of requests
-const _limit = 3 // limit for requests within interval
+const _interval = 10000 // milliseconds to check number of requests
+const _limit = 10 // limit for requests within interval
+let totalActiveUsers = 0 // The amount of active users reset every now and then (default 30 sec)
```
[the remark this answers is not in the document]
Comment (Samuelhertzberg) [reply index.js]: Yes, my mistake

File: index.js (hunk 1 above)
Comment (Samuelhertzberg) [reply index.js]: Yes, we will change to a flat config structure and check that userActiveTimeout is a number

File: index.js (hunk 2 above)
Comment (Samuelhertzberg) [reply index.js]: When demonstrating, 10 seconds is a good time span considering that several users will have to connect to the server.

File: index.js (hunk 3 above)
Comment (Samuelhertzberg) [reply index.js]: Yes, but we will have to add it during demonstration.

File: index.js [hunk 5]
```diff
@@ -76,7 +91,22 @@ const dostroy = (config) => async (req, res, next) => {
   const all = !config || Object.keys(config).length === 0
   const r = config && config.rudy ? config.rudy : RUDY_DEFAULT
   const rl = config && config.rateLimiting ? config.rateLimiting : RATELIMITING_DEFAULT
+  const dynamic = config && config.dynamicRateLimiting ? config.dynamicRateLimiting : USE_DYNAMIC_RATE_LIMITING_DEFAULT
+  const userActiveTimeout = dynamic && config.userActiveTimeout && !isNaN(config.userActiveTimeout) ? config.userActiveTimeout : USER_ACTIVE_TIMEOUT_DEFAULT
```
Comment (helenaalinder) [inline index.js]: If config === undefined this will throw exception. consider: `dynamic && config && !isNaN(config.userActiveTimeout)`

Comment (helenaalinder) [reply index.js]: isNaN handles undefined values

Comment (Samuelhertzberg) [reply index.js]: Yes, this is true. If config is null this will crash :O

**Pull request merged**